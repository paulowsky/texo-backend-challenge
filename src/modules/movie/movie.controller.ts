import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  Param,
  Patch,
  Post
} from '@nestjs/common'

import { CreateMovieInput, UpdateMovieInput } from './dto/movie.input'
import { Movie } from './entities/movie.entity'
import { MovieService } from './movie.service'

@Controller('movie')
export class MovieController {
  constructor(private movieService: MovieService) {}

  @HttpCode(200)
  @Get('')
  async getAll(): Promise<Movie[]> {
    return this.movieService.getAll()
  }

  @HttpCode(200)
  @Get(':title')
  async getOne(@Param('title') title: string): Promise<Movie> {
    return this.movieService.getOne(title)
  }

  @HttpCode(201)
  @Post('')
  async create(@Body() data: CreateMovieInput): Promise<Movie> {
    return this.movieService.create(data)
  }

  @HttpCode(200)
  @Patch('')
  async update(@Body() data: UpdateMovieInput): Promise<Movie> {
    return this.movieService.update(data)
  }

  @HttpCode(200)
  @Delete(':title')
  async delete(@Param('title') title: string): Promise<void> {
    return this.movieService.delete(title)
  }
}
