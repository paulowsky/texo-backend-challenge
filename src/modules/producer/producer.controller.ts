import { Controller, Get, HttpCode } from '@nestjs/common'

import { Intervals } from './entities/intervals.entity'

import { ProducerService } from './producer.service'

@Controller('producer')
export class ProducerController {
  constructor(private producerService: ProducerService) {}

  @HttpCode(200)
  @Get('intervals')
  async intervals(): Promise<Intervals> {
    return this.producerService.getIntervals()
  }
}
