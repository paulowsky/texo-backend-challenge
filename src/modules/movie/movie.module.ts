import { Module } from '@nestjs/common'

import { DatabaseModule } from '@db/db.module'

import { MovieController } from './movie.controller'
import { MovieService } from './movie.service'

@Module({
  imports: [DatabaseModule],
  controllers: [MovieController],
  providers: [MovieService],
  exports: [MovieService]
})
export class MovieModule {}
