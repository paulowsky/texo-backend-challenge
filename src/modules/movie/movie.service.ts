import { Injectable, NotFoundException } from '@nestjs/common'

import { DatabaseMovie, DatabaseService } from '@db/db.service'

import { CreateMovieInput, UpdateMovieInput } from './dto/movie.input'

@Injectable()
export class MovieService {
  constructor(private readonly db: DatabaseService) {}

  async getAll(filter?: Partial<DatabaseMovie>) {
    const result = await this.db.findAllMovies(filter)

    const movies = result.map(movie => ({
      ...movie,
      year: parseInt(movie.year),
      studios: movie.studios.split(', ').map(str => str.trim()),
      producers: movie.producers
        .replace(' and ', ', ')
        .split(', ')
        .map(str => str.trim()),
      winner: movie.winner === 'yes' ? true : false
    }))

    return movies
  }

  async getOne(title: string) {
    const result = await this.getAll({ title })

    if (!result.length) throw new NotFoundException('Movie not found')

    return result[0]
  }

  async create(data: CreateMovieInput) {
    const normalizedData = {
      title: data.title,
      winner: data.winner ? 'yes' : '',
      year: data.year.toString(),
      studios: data.studios.join(', '),
      producers: data.producers.join(', ')
    }

    await this.db.createMovie(normalizedData)

    return await this.getOne(normalizedData.title)
  }

  async update(data: UpdateMovieInput) {
    await this.getOne(data.originalTitle)

    const normalizedData = {
      title: data.title,
      winner: data.winner ? 'yes' : '',
      year: data.year.toString(),
      studios: data.studios.join(', '),
      producers: data.producers.join(', ')
    }

    await this.db.updateMovie(data.originalTitle, normalizedData)

    return await this.getOne(
      normalizedData.title ? normalizedData.title : data.originalTitle
    )
  }

  async delete(title: string) {
    await this.getOne(title)

    await this.db.deleteMovie(title)
  }
}
