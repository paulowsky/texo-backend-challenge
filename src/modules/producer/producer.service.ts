import { Injectable } from '@nestjs/common'

import { DatabaseService } from '@db/db.service'

import { Movie } from '@modules/movie/entities/movie.entity'
import { MovieService } from '@modules/movie/movie.service'

@Injectable()
export class ProducerService {
  constructor(
    private readonly db: DatabaseService,
    private readonly movieService: MovieService
  ) {}

  async getAll(): Promise<string[]> {
    const movies = await this.movieService.getAll()

    return Array.from(
      new Set(movies.reduce((acc, winner) => acc.concat(winner.producers), []))
    )
  }

  async getIntervals() {
    const movies = await this.movieService.getAll({ winner: 'yes' })

    const producerIntervals = await this.selectEligibleProducers(movies)

    const min = this.getMinIntervals(producerIntervals)

    const max = this.getMaxIntervals(producerIntervals)

    return { min, max }
  }

  getMinIntervals(
    producerIntervals: {
      producer: string
      minInterval: {
        value: number
        previousWin: number
        followingWin: number
      }
      maxInterval: {
        value: number
        previousWin: number
        followingWin: number
      }
    }[]
  ) {
    let min = []

    producerIntervals.forEach(interval => {
      const parsedInterval = {
        producer: interval.producer,
        interval: interval.minInterval.value,
        previousWin: interval.minInterval.previousWin,
        followingWin: interval.minInterval.followingWin
      }

      if (!min.length || parsedInterval.interval < min[0].interval) {
        min = [parsedInterval]
      } else if (parsedInterval.interval === min[0].interval) {
        min.push(parsedInterval)
      }
    })

    return min
  }

  getMaxIntervals(
    producerIntervals: {
      producer: string
      minInterval: {
        value: number
        previousWin: number
        followingWin: number
      }
      maxInterval: {
        value: number
        previousWin: number
        followingWin: number
      }
    }[]
  ) {
    let max = []

    producerIntervals.forEach(interval => {
      const parsedInterval = {
        producer: interval.producer,
        interval: interval.maxInterval.value,
        previousWin: interval.maxInterval.previousWin,
        followingWin: interval.maxInterval.followingWin
      }

      if (!max.length || parsedInterval.interval > max[0].interval) {
        max = [parsedInterval]
      } else if (parsedInterval.interval === max[0].interval) {
        max.push(parsedInterval)
      }
    })

    return max
  }

  async selectEligibleProducers(movies: Movie[]): Promise<
    {
      producer: string
      minInterval: {
        value: number
        previousWin: number
        followingWin: number
      }
      maxInterval: {
        value: number
        previousWin: number
        followingWin: number
      }
    }[]
  > {
    const producers = await this.getAll()

    const eligible = []

    producers.forEach(producer => {
      const producerMovies = movies.filter(movie =>
        movie.producers.includes(producer)
      )

      if (producerMovies.length > 1) {
        let minInterval: {
          value: number
          previousWin: number
          followingWin: number
        }

        let maxInterval: {
          value: number
          previousWin: number
          followingWin: number
        }

        for (let i = 1; i < producerMovies.length; i++) {
          const interval = {
            value: producerMovies[i].year - producerMovies[i - 1].year,
            previousWin: producerMovies[i - 1].year,
            followingWin: producerMovies[i].year
          }

          if (!minInterval || interval.value < minInterval.value)
            minInterval = interval

          if (!maxInterval || interval.value > maxInterval.value)
            maxInterval = interval
        }

        eligible.push({ producer, minInterval, maxInterval })
      }
    })

    return eligible
  }
}
