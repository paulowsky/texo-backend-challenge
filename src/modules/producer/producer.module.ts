import { Module } from '@nestjs/common'

import { DatabaseModule } from '@db/db.module'

import { MovieModule } from '@modules/movie/movie.module'

import { ProducerController } from './producer.controller'
import { ProducerService } from './producer.service'

@Module({
  imports: [DatabaseModule, MovieModule],
  controllers: [ProducerController],
  providers: [ProducerService],
  exports: [ProducerService]
})
export class ProducerModule {}
