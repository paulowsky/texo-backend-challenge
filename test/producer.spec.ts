import { INestApplication } from '@nestjs/common'
import { Test } from '@nestjs/testing'
import * as request from 'supertest'

import { ProducerController } from '@modules/producer/producer.controller'
import { ProducerService } from '@modules/producer/producer.service'
import { MovieService } from '@modules/movie/movie.service'

import { mockMovieService, mockProducerService } from './fixtures/mocks'

describe('Movie Module', () => {
  let app: INestApplication

  beforeAll(async () => {
    const moduleFixture = await Test.createTestingModule({
      controllers: [ProducerController],
      providers: [
        {
          provide: MovieService,
          useValue: mockMovieService
        },
        {
          provide: ProducerService,
          useValue: mockProducerService
        }
      ]
    }).compile()

    app = moduleFixture.createNestApplication()

    await app.init()
  })

  afterAll(async () => {
    await app.close()
  })

  it('(GET) /producer/intervals', async () => {
    const response = await request(app.getHttpServer()).get(
      '/producer/intervals'
    )

    expect(response.statusCode).toEqual(200)
    expect(response.body).toMatchSnapshot()
  })
})
