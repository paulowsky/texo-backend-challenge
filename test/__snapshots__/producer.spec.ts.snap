// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Module (GET) /producer/intervals 1`] = `
[
  {
    "max": [],
    "min": [],
  },
]
`;
